# Table of Contents
1. Extract raw data
2. 
3. 

In [1]:
import pandas as pd
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)
import xml.etree.ElementTree as ET
import json
import os
import string

## 1. Extract raw data
Prepare a pandas dataframe containing all of the raw data provided for the exercise 

In [2]:
data_path = 'C:\\Users\\Shreye\\Documents\\data-scientist-exercise02\\data\\'

In [3]:
# Load XML file into dataframe
# https://stackoverflow.com/questions/41795198/more-efficient-conversion-of-xml-file-into-dataframe.
etree = ET.parse(data_path + 'AviationData.xml') 
data = []
for el in etree.iterfind('./*'):
    for i in el.iterfind('*'):
        data.append(dict(i.items()))       
aviation = pd.DataFrame(data)
aviation.shape

(77257, 31)

In [4]:
# Load JSON files into dataframe
narrative = pd.DataFrame()
all_jsons = [file for file in os.listdir(data_path) if file!='AviationData.xml'] # get a list of all of the JSON file names
for json_name in all_jsons:
    with open(data_path+json_name) as json_file: # read each JSON into a dataframe
        json_content = json_file.read()
        json_data = json.loads(json_content)
        json_df = pd.DataFrame(json_data['data']) # the "dataframe" is within the 'data' field of the JSON file
        narrative = pd.concat([narrative, json_df]) # concatenate the JSON dataframes into one
narrative.shape

(76133, 3)

In [5]:
# Merge the dataframes on the common EventId field
ntsb = aviation.merge(narrative, how='left')
print(ntsb.shape)
ntsb.describe()

(77257, 33)


AccidentNumber AirCarrier AircraftCategory AircraftDamage AirportCode AirportName AmateurBuilt BroadPhaseOfFlight        Country     EngineType   EventDate         EventId FARDescription InjurySeverity InvestigationType Latitude       Location Longitude    Make  Model NumberOfEngines PublicationDate PurposeOfFlight RegistrationNumber    ReportStatus Schedule TotalFatalInjuries TotalMinorInjuries TotalSeriousInjuries TotalUninjured WeatherCondition narrative probable_cause
count           77257      77257            77257          77257       77257       77257        77257              77257          77257          77257       77257           77257          77257          77257             77257    77257          77257     77257   77257  77257           77257           77257           77257              77257           77257    77257              77257              77257                77257          77257            77257     77257          77257
unique          77257       2814               13              4        9489       22284            3                 13            174             15       12180           76133             17            120                 2    16343          24702     17471    7204  11029               8            3403              23              67493               4        4                118                 63                   41            364                4     75557          47435
top        WPR11LA263                                Substantial                                   No            LANDING  United States  Reciprocating  05/16/1982  20100204X45658                     Non-Fatal          Accident           ANCHORAGE, AK            CESSNA    152               1                        Personal                     Probable Cause                           0                  0                    0              1              VMC                         
freq                1      73439            60737          55420       33780       29926        69198              18553          73076          63016          25               3          60592          58499             74207    53496            372     53505   16609   2251           61465           13188           43360               2756           72264    65878              40363              40342                42955          22029            68764       231          26754

> From all accidents, the most commmon phase of flight is LANDING and the most common aircraft is a Cessna 152

In [ ]:
# ntsb.replace(['Unavailable', 'UNK', 'UNKNOWN', 'Unknown', ], '', inplace=True)

# ntsb.InjurySeverity = ntsb.InjurySeverity.apply(lambda x: x.translate(str.maketrans('','',string.punctuation)))
# ntsb.InjurySeverity = ntsb.InjurySeverity.apply(lambda x: x.translate(str.maketrans('','',string.digits)))